tests/system/test_dbapi.py

```diff
@@ -344,6 +343,133 @@ def test_DDL_autocommit(shared_instance, dbapi_database):
             op.result()
 
 
+def test_ddl_execute_autocommit_true(shared_instance, dbapi_database):
+    """Check that DDL statement in autocommit mode results in successful
+    DDL statement execution for execute method."""
+
+    conn = Connection(shared_instance, dbapi_database)
+    conn.autocommit = True
+    cur = conn.cursor()
+    cur.execute(
+        """
+        CREATE TABLE DdlExecuteAutocommit (
+            SingerId     INT64 NOT NULL,
+            Name    STRING(1024),
+        ) PRIMARY KEY (SingerId)
+    """
+    )
+    table = dbapi_database.table("DdlExecuteAutocommit")
+    assert table.exists() is True
+
+    cur.close()
+    conn.close()
+
+
+def test_ddl_executemany_autocommit_true(shared_instance, dbapi_database):
+    """Check that DDL statement in autocommit mode results in exception for
+    executemany method ."""
+
+    conn = Connection(shared_instance, dbapi_database)
+    conn.autocommit = True
+    cur = conn.cursor()
+    with pytest.raises(ProgrammingError):
+        cur.executemany(
+            """
+            CREATE TABLE DdlExecuteManyAutocommit (
+                SingerId     INT64 NOT NULL,
+                Name    STRING(1024),
+            ) PRIMARY KEY (SingerId)
+        """,
+            [],
+        )
+    table = dbapi_database.table("DdlExecuteManyAutocommit")
+    assert table.exists() is False
+
+    cur.close()
+    conn.close()
+
+
+def test_ddl_execute(shared_instance, dbapi_database):
+    """Check that DDL statement followed by non-DDL execute statement in
+    non autocommit mode results in successful DDL statement execution."""
+
+    conn = Connection(shared_instance, dbapi_database)
+    want_row = (
+        1,
+        "first-name",
+    )
+    cur = conn.cursor()
+    cur.execute(
+        """
+        CREATE TABLE DdlExecute (
+            SingerId     INT64 NOT NULL,
+            Name    STRING(1024),
+        ) PRIMARY KEY (SingerId)
+    """
+    )
+    table = dbapi_database.table("DdlExecute")
+    assert table.exists() is False
+
+    cur.execute(
```

Comment:
This behavior is what has been implemented, but it is really weird. For an outside observer (and especially one that is not familiar with Cloud Spanner and/or this dbapi driver), it will seem like:
1. Cloud Spanner is super fast in executing DDL statements. They finish in like 10 microseconds!
2. Cloud Spanner is super slow in executing DML statements. They take like 10 seconds!

Would you mind writing a small one-pager on this behavior and share it with the rest of the Connectivity team + Adam + .... (maybe more)? I think that we need to discuss whether we want to keep this behavior, or otherwise document it very well. I'm surprised that we haven't had any questions regarding this yet.

Reply:
Sure, will start a doc